src/wp-includes/html-api/class-wp-html-processor.php

```diff
@@ -4113,9 +4113,26 @@ private function step_in_foreign_content(): bool {
 				return true;
 
 			/*
-			 * > A comment token
+			 * A CDATA section is effectively a text node and should have similar handling.
 			 */
 			case '#cdata-section':
+				/*
+				 * NULL bytes and whitespace do not change the frameset-ok flag.
+				 */
+				$current_token        = $this->bookmarks[ $this->state->current_token->bookmark_name ];
```

Comment:
I get what we're doing here, but only after a few stages of review. what changes if we call `get_modifiable_text()` here?

Reply:
`get_modifiable_text` will replace the null byte with the replacement character, so we need to inspect the raw text for null bytes. The spec is clear, at a null byte the replacement character is inserted but it does not change the frameset-ok flag.

Reply:
as CDATA I suppose that there's no possible way we would need to decode character references, so this seems okay. I don't know how I feel about the bookmark trick, but I've wrestled this this many times (though I don't think I've yet needed a kind of `get_raw_text()`)

Reply:
this makes me go back and rethink the null byte substitution. there are definitely nice things about leaving them in through the levels

Reply:
@sirreal the more I consider this the less I like these special-casings for text stuff. I want to try exploring the idea we discussed, which was adding a kind of `subdivide_text_node_appropriately()` method that would split text nodes into `NULL` byte sequences and other text.

the one pitfall of this method is handling whitespace, since most HTML insertion modes consider the _character token_ and not the raw bytes alone, while for segments like CDATA these are the same.

still, I think there are potential performance wins there, as we can skip decoding character references if we know, for example, that there are no `&`s in there. I'm going to create another PR to try this, and if it works, I think it will improve all of the text-checking rules and speed up parsing

Reply:
okay now I know one reason I don't like this code with the bookmark - I've seen it and written it before 😆

well in #7236 I've explored the idea of subdividing text appropriately and feel very good about it. I think that overall it allows us to defer some amount of text decoding, and the logic seems clearer to me.

surprisingly the change leads to four more tests being skipped - I'm not sure why and whether that's proper or not.

Reply:
oops - I think that PR solves the problem this is reporting and fixing. so it's an alternate approach I guess, but one that does more.

Reply:
Since #7236 doesn't fix the CDATA nodes (https://github.com/WordPress/wordpress-develop/pull/7236#issuecomment-2310812909), there may be a place for both changes.

Reply:
We agreed via another channel that both #7236 and this change have a place, where #7236 handles text nodes more generally.